src/blog/short-builtin-calls.md

```diff
@@ -0,0 +1,45 @@
+avatars:
+  - toon-verwaest
+date: 2021-05-06
+tags:
+  - JavaScript
+description: 'In V8 9.1 we’ve temporarily unembedded the builtins on desktop to avoid performance issues resulting from far indirect calls.'
+tweet: ''
+---
+
+In V8 9.1 we’ve temporarily disabled [embedded builtins](https://v8.dev/blog/embedded-builtins) on desktop. While embedding builtins significantly improves memory usage, we’ve realized that function calls between embedded builtins and JIT compiled code can come at a considerable performance penalty. This cost depends on the microarchitecture of the CPU. In this post we’ll explain why this is happening, what the performance looks like, and what we’re planning to do to resolve this long-term.
+
+## Code Allocation
+
+Machine code generated by V8’s just-in-time (JIT) compilers is allocated dynamically on memory pages owned by the VM. V8 allocates memory pages within a contiguous address space region, which itself either lies somewhere randomly in memory (for [address space layout randomization](https://en.wikipedia.org/wiki/Address_space_layout_randomization) reasons), or somewhere inside of the 4gib virtual memory cage we allocate for [pointer compression](https://v8.dev/blog/pointer-compression).
+
+V8 JIT code very commonly calls into builtins. Builtins are essentially snippets of machine code that are shipped as part of the VM. There are builtins that implement full JavaScript standard library functions, such as `Function.prototype.bind`, but many builtins are helper snippets of machine code that fill in the gap between the higher-level semantics of JS and the low-level capabilities of the CPU. For example, if a JavaScript function wants to call another JavaScript function, it is common for the function implementation to call a `CallFunction` builtin that figures out how the target JavaScript function should be called; i.e., whether it’s a proxy or a regular function, how many arguments it expects, etc. Since these snippets are known when we build the VM, they are "embedded" in the Chrome binary, which means that they end up within the Chrome binary code region.
+
+## Direct vs Indirect Calls
+
+On 64 bit architectures, the Chrome binary, which includes these builtins, lies arbitrarily far away from JIT code. With the [x86-64](https://en.wikipedia.org/wiki/X86-64) instruction set, this means we can’t use direct calls: they take a 32-bit signed immediate that’s used as an offset to the address of the call, and the target may be more than 2 GiB away. Instead, we need to rely on indirect calls through a register or memory operand. Such calls rely more heavily on prediction since it’s not immediately apparent from the call instruction itself what the target of the call is. On [ARM64](https://en.wikipedia.org/wiki/AArch64) we can’t use direct calls at all since the range is limited to 128 MiB. This means that in both cases we rely on the accuracy of the CPU's indirect branch predictor.
```

Comment:
```suggestion
On 64-bit architectures, the Chrome binary, which includes these builtins, lies arbitrarily far away from JIT code. With the [x86-64](https://en.wikipedia.org/wiki/X86-64) instruction set, this means we can’t use direct calls: they take a 32-bit signed immediate that’s used as an offset to the address of the call, and the target may be more than 2 GiB away. Instead, we need to rely on indirect calls through a register or memory operand. Such calls rely more heavily on prediction since it’s not immediately apparent from the call instruction itself what the target of the call is. On [ARM64](https://en.wikipedia.org/wiki/AArch64) we can’t use direct calls at all since the range is limited to 128 MiB. This means that in both cases we rely on the accuracy of the CPU's indirect branch predictor.
```